File: src/p1171/java/solution.java
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        ListNode sentinel = new ListNode(0);
        sentinel.next = head;
        HashMap<Integer, ListNode> last = new HashMap<>();
        int total = 0;
        for (ListNode p = sentinel; p != null; p = p.next) {
            total += p.val;
            last.put(total, p);
        }
        total = 0;
        for (ListNode p = sentinel; p != null; p = p.next) {
            total += p.val;
            p.next = last.get(total).next;
        }
        return sentinel.next;
    }
}

File: src/p1171/golang/solution.go
func removeZeroSumSublists(head *ListNode) *ListNode {
	sentinel := ListNode{Next: head}
	last := make(map[int]*ListNode)
	total := 0
	for p := &sentinel; p != nil; p = p.Next {
		total += p.Val
		last[total] = p
	}
	total = 0
	for p := &sentinel; p != nil; p = p.Next {
		total += p.Val
		p.Next = last[total].Next
	}
	return sentinel.Next
}
